### PR TITLE
Add PyTorch example notebook

### PR DESCRIPTION
**Pull request by mkuchnik**
Add an end-to-end example of how to use Croissant in PyTorch using an adapter on top of mlcroissant and FLORES-200. The goal of the notebook is to provide a minimal yet somewhat realistic demo of how Croissant can be used in training. I am planning to expand on some of the features (e.g., performance) in future patches.

### CODE REVIEW

**Comment by mkuchnik** [comment]
The 3 CI issues seem non-blocking to me, but let me know if you think otherwise:

* Python format error is outside this patch (log says `croissant/croissant/editor/events/resources.py`)
* The 2 Notebook tests are 10 minute timeouts likely due to ML training. The training time can be shortened to pass CI (and it would indeed be nice to tune the notebook to be faster), but I'd prefer to address that in a later patch. The notebook state already shows the full run, so run latency should not be a blocking factor.

File: python/mlcroissant/recipes/torch_adapter/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        data_type = infer_data_type(val)
+        return apply_data_type_transformation(val, data_type)
+
+    if data_type == LoaderSpecificationDataType.UTF8:
+        return val.decode("utf-8")
+
+    return val
+
+
+class LoaderFactory:
```
Comment (marcenacp) [inline python/mlcroissant/recipes/torch_adapter/dataloader.py:49]: Do you need unit tests for this?

**Review by marcenacp — CHANGES_REQUESTED** [review]
This is a great addition. Thanks for taking the time to implement this, Michael!

**Comment by mkuchnik** [comment]
@marcenacp What are your thoughts on external dependencies? I attempted to make `torchdata` optional in incorporating the changes to `mlcroissant`, but I'm not sure how to handle this in the CI, which either does or does not need the dependencies. Currently, the PyType test is failing because of the dynamic import behavior of optional dependencies.

File: python/mlcroissant/recipes/PyTorch_FLORES200.ipynb [hunk 2]
```diff
@@ -0,0 +1,757 @@
+{
```
[the remark this answers is not in the document]
Comment (mkuchnik) [reply python/mlcroissant/recipes/PyTorch_FLORES200.ipynb:1]: Should be updated now

File: python/mlcroissant/recipes/torch_adapter/dataloader.py (hunk 1 above)
Comment (mkuchnik) [reply python/mlcroissant/recipes/torch_adapter/dataloader.py]: Added unit tests

File: python/mlcroissant/recipes/PyTorch_FLORES200.ipynb [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import functools\n",
+    "import time\n",
+    "\n",
+    "import numpy as np\n",
+    "import torch\n",
+    "from torch.utils.data import DataLoader\n",
+    "import torch_adapter as ta\n",
```
[the remark this answers is not in the document]
Comment (mkuchnik) [reply python/mlcroissant/recipes/PyTorch_FLORES200.ipynb]: Take a look at the current patch. I tried to make `mlcroissant.torch` have optional dependencies, but it makes the CI complicated.

**Comment by mkuchnik** [comment]
@marcenacp Current patch passes CI and keeps framework dependencies as optional until used. Let me know if you have suggestions!

**Review by marcenacp — APPROVED** [review]
Thanks! Sorry for the delay in the review. Also left a few minor comments :)

**Comment by mkuchnik** [comment]
@marcenacp I took my pass over the suggestions. Let me know if there are any outstanding issues.

**Review by marcenacp — APPROVED** [review]
Thanks!

**Pull request merged**